component/admin/controllers/stockroom.php

```diff
@@ -9,90 +9,73 @@
 {
-    public function cancel()
-    {
-        $this->setRedirect('index.php');
-    }
-
-    public function listing()
-    {
-        $this->setRedirect('index.php?option=com_redshop&view=stockroom_listing&id=0');
-    }
-
-    public function publish()
-    {
-        $cid = $this->input->post->get('cid', array(0), 'array');
-
-        if (!is_array($cid) || count($cid) < 1) {
-            throw new Exception(JText::_('COM_REDSHOP_SELECT_AN_ITEM_TO_PUBLISH'));
-        }
-
-        /** @var RedshopModelStockroom_detail $model */
-        $model = $this->getModel('stockroom_detail');
-
-        if (!$model->publish($cid, 1)) {
-            echo "<script> alert('" . $model->getError(true) . "'); window.history.go(-1); </script>\n";
-        }
-
-        $msg = JText::_('COM_REDSHOP_STOCK_ROOM_DETAIL_PUBLISHED_SUCCESSFULLY');
-        $this->setRedirect('index.php?option=com_redshop&view=stockroom', $msg);
-    }
-
-    public function unpublish()
-    {
-        $cid = $this->input->post->get('cid', array(0), 'array');
-
-        if (!is_array($cid) || count($cid) < 1) {
-            throw new Exception(JText::_('COM_REDSHOP_SELECT_AN_ITEM_TO_UNPUBLISH'));
-        }
-
-        /** @var RedshopModelStockroom_detail $model */
-        $model = $this->getModel('stockroom_detail');
-
-        if (!$model->publish($cid, 0)) {
-            echo "<script> alert('" . $model->getError(true) . "'); window.history.go(-1); </script>\n";
-        }
-
-        $msg = JText::_('COM_REDSHOP_STOCK_ROOM_DETAIL_UNPUBLISHED_SUCCESSFULLY');
-        $this->setRedirect('index.php?option=com_redshop&view=stockroom', $msg);
-    }
-
-    public function frontpublish()
-    {
-        $cid = $this->input->post->get('cid', array(0), 'array');
-
-        if (!is_array($cid) || count($cid) < 1) {
-            throw new Exception(JText::_('COM_REDSHOP_SELECT_AN_ITEM_TO_PUBLISH'));
-        }
-
-        $model = $this->getModel('stockroom_detail');
-
-        if (!$model->frontpublish($cid, 1)) {
-            echo "<script> alert('" . $model->getError(true) . "'); window.history.go(-1); </script>\n";
-        }
-
-        $msg = JText::_('COM_REDSHOP_STOCK_ROOM_DETAIL_PUBLISHED_SUCCESSFULLY');
-        $this->setRedirect('index.php?option=com_redshop&view=stockroom', $msg);
-    }
-
-    public function frontunpublish()
-    {
-        $cid = $this->input->post->get('cid', array(0), 'array');
-
-        if (!is_array($cid) || count($cid) < 1) {
-            throw new Exception(JText::_('COM_REDSHOP_SELECT_AN_ITEM_TO_UNPUBLISH'));
-        }
-
-        $model = $this->getModel('stockroom_detail');
-
-        if (!$model->frontpublish($cid, 0)) {
-            echo "<script> alert('" . $model->getError(true) . "'); window.history.go(-1); </script>\n";
-        }
-
-        $msg = JText::_('COM_REDSHOP_STOCK_ROOM_DETAIL_UNPUBLISHED_SUCCESSFULLY');
-        $this->setRedirect('index.php?option=com_redshop&view=stockroom', $msg);
-    }
```

Comment:
@hoang-tran-redweb i think you should check this one because when removing 2 methods above the button show in frontend will not work

Reply:
I fixed.